sdks/java/extensions/avro/src/test/java/org/apache/beam/sdk/extensions/avro/coders/AvroCoderTest.java

```diff
@@ -323,35 +314,71 @@ public void testKryoSerialization() throws Exception {
   @Test
   public void testPojoEncoding() throws Exception {
     Pojo value = new Pojo("Hello", 42, DATETIME_A);
-    AvroCoder<Pojo> coder = AvroCoder.of(Pojo.class);
+    AvroCoder<Pojo> coder = AvroCoder.reflect(Pojo.class);
 
     CoderProperties.coderDecodeEncodeEqual(coder, value);
   }
 
   @Test
   public void testSpecificRecordEncoding() throws Exception {
-    if (isBrokenMapComparison()) {
-      // Don't compare the map values because of AVRO-2943
-      AVRO_SPECIFIC_RECORD.setMap(ImmutableMap.of());
-    }
-    AvroCoder<TestAvro> coder =
-        AvroCoder.of(TestAvro.class, AVRO_SPECIFIC_RECORD.getSchema(), false);
+    // Don't compare the map values because of AVRO-2943
+    AVRO_SPECIFIC_RECORD.setMap(ImmutableMap.of());
+
+    AvroCoder<TestAvro> coder = AvroCoder.specific(TestAvro.class);
+    AvroCoder<TestAvro> coderWithSchema =
+        AvroCoder.specific(TestAvro.class, AVRO_SPECIFIC_RECORD.getSchema());
 
     assertTrue(SpecificRecord.class.isAssignableFrom(coder.getType()));
+    assertTrue(SpecificRecord.class.isAssignableFrom(coderWithSchema.getType()));
+
     CoderProperties.coderDecodeEncodeEqual(coder, AVRO_SPECIFIC_RECORD);
+    CoderProperties.coderDecodeEncodeEqual(coderWithSchema, AVRO_SPECIFIC_RECORD);
   }
 
-  private boolean isBrokenMapComparison() {
-    return VERSION_AVRO.equals("1.9.2")
-        || VERSION_AVRO.equals("1.10.2")
-        || VERSION_AVRO.equals("1.11.1");
+  // example to overcome AVRO-2943 limitation with custom datum factory
+  // force usage of String instead of Utf8 when avro type is set to CharSequence
+  static class CustomSpecificDatumFactory<T> extends AvroDatumFactory.SpecificDatumFactory<T> {
```

Comment:
@mosche here is an example of a custom `AvroDatumFactory`.
In this case, we change the default model creation by using `String` as underlying `CharSequence` implementation instead of `Utf8`